src/esp/sensor/CameraSensor.cpp

```diff
@@ -150,6 +140,11 @@ bool CameraSensor::drawObservation(sim::Simulator& sim) {
   if (sim.isFrustumCullingEnabled())
     flags |= gfx::RenderCamera::Flag::FrustumCulling;
 
+  // Set projection matrix, transformation matrix, and viewport
+  renderCamera_->setProjectionMatrix(width_, height_, projectionMatrix_);
```

Comment:
You do not need to set it every time.
You can consider **setting it once if there is any changes** to the projectionMatrix_.
- So you will set it first in the constructor;
- Then you will set it when `projectionMatrix_` needs to be updated. Note that `projectionMatrix_` is recomputed in the function `recomputeProjectionMatrix()`, which means you can call `renderCamera_->setProjectionMatrix(...)` there. 

Careful: after you make the above changes, re-visit the `viewportEvent()` in the viewer.cpp. That function handles the case where the window size is changed (e.g., you use your mouse to enlarge the window). You have to make sure changes on width and height can eventually trigger the re-computation of the projection matrix!!